lib/Service/TokenService.php

```diff
@@ -0,0 +1,293 @@
+	 * @throws DoesNotExistException
+	 * @throws MultipleObjectsReturnedException
+	 * @throws TokenExchangeFailedException
+	 * @throws \JsonException
+	 */
+	public function getExchangedToken(string $targetAudience): Token {
+		$loginToken = $this->getToken();
+		if ($loginToken === null) {
+			$this->logger->debug('[TokenService] Failed to exchange token, no login token found in the session');
+			throw new TokenExchangeFailedException('Failed to exchange token, no login token found in the session');
+		}
+		if ($loginToken->isExpired()) {
+			$this->logger->debug('[TokenService] Failed to exchange token, the login token is expired');
+			throw new TokenExchangeFailedException('Failed to exchange token, the login token is expired');
+		}
+		$oidcProvider = $this->providerMapper->getProvider($loginToken->getProviderId());
+		$discovery = $this->discoveryService->obtainDiscovery($oidcProvider);
+
+		try {
+			$clientSecret = $oidcProvider->getClientSecret();
+			if ($clientSecret !== '') {
+				try {
+					$clientSecret = $this->crypto->decrypt($clientSecret);
+				} catch (\Exception $e) {
+					$this->logger->error('[TokenService] Token Exchange: Failed to decrypt oidc client secret');
+				}
+			}
+			$this->logger->debug('[TokenService] Exchanging the token: ' . $discovery['token_endpoint']);
+			// more in https://www.keycloak.org/securing-apps/token-exchange
+			$result = $this->client->post(
+				$discovery['token_endpoint'],
+				[
+					'body' => [
+						'client_id' => $oidcProvider->getClientId(),
+						'client_secret' => $clientSecret,
+						'grant_type' => 'urn:ietf:params:oauth:grant-type:token-exchange',
+						'subject_token' => $loginToken->getAccessToken(),
+						'subject_token_type' => 'urn:ietf:params:oauth:token-type:access_token',
+						// can also be
+						// urn:ietf:params:oauth:token-type:access_token
+						// or urn:ietf:params:oauth:token-type:id_token
+						// this one will get us an access token and refresh token within the response
+						'requested_token_type' => 'urn:ietf:params:oauth:token-type:refresh_token',
+						'audience' => $targetAudience,
+					],
+				]
+			);
+			$this->logger->debug('[TokenService] Token exchange request params', [
+				'client_id' => $oidcProvider->getClientId(),
+				// 'client_secret' => $clientSecret,
+				'grant_type' => 'urn:ietf:params:oauth:grant-type:token-exchange',
+				// 'subject_token' => $loginToken->getAccessToken(),
+				'subject_token_type' => 'urn:ietf:params:oauth:token-type:access_token',
+				'requested_token_type' => 'urn:ietf:params:oauth:token-type:refresh_token',
+				'audience' => $targetAudience,
+			]);
+			$body = $result->getBody();
+			$bodyArray = json_decode(trim($body), true, 512, JSON_THROW_ON_ERROR);
+			$this->logger->debug('[TokenService] Token exchange success: "' . trim($body) . '"');
+			$tokenData = array_merge(
+				$bodyArray,
+				['provider_id' => $loginToken->getProviderId()],
+			);
+			return new Token($tokenData);
```

Comment:
Hmmm in my tests the exchanged token has the same "shape" as the login token, it contains an id token and an access token. So it might depend on the client config on the IdP side. We can make it so the Token object does not require an id_token.

Reply:
The id_token might be used for anything, not just the login. It contains user info indeed.

Reply:
Can you try to change line 41 of `lib/Model/Token.php` to:
`$this->idToken = $tokenData['id_token'] ?? null;`

Reply:
> Hmmm in my tests the exchanged token has the same "shape" as the login token, it contains an id token and an access token. So it might depend on the client config on the IdP side. We can make it so the Token object does not require an id_token.
Sure, I think we do not require the id token to be stored for client to client communication for oidc based authentication.

Reply:
and 32 to:
`private ?string $idToken;`

Reply:
yes i did and its working!